### DISCUSSION TITLE
System Benchmarks

### DISCUSSION THREAD

**Discussion by JaminMartin** [ankrh/MCmatlab#3] [Q&A]
Hi, 

I wondered if anyone has run some benchmarking tests for the recently compiled MacOS binaries. I am considering an apple silicon device, but it would be good to get an idea of the performance compared to my current devices. 

Perhaps a benchmark of example 16 (CUDA) could be used. I currently have a desktop running Linux & Windows with a CUDA-capable card & a thin and light laptop to which I can contribute data. I imagine a decent runtime would be required to see if there are any thermal limits. 

This is an amazing project and would like to keep using it and hopefully contribute to it in the future. Thank you for the effort that has gone into this so far.

**Comment by DrDonik** [comment]
Hi Sacroth,

we haven't done any proper benchmarking yet. Part of the reason is that we don't have access to a Mac with a CUDA-capable graphics card nor to an Apple silicon device. So we have never tried to compile the CUDA-version of the binaries for macOS, and, by now (2022), NVIDIA even has discontinued CUDA support for macOS, with the last supported version being 10.2 from 2018. Even on non-cuda hardware, another challenge is that the native macOS compiler for MATLAB doesn't support openmp and, thus, multithreading is not natively possible. It can be done on intel macs by compiling the binaries with llvm from homebrew (as is done for the shipped binaries that will use multithreading), but I don't know how or if at all this works on Apple silicon.

In short: I would need to get my hands on an Apple silicon device to be able to supply the data you seek.

I agree both would be something I'm interested in (getting an Apple silicon device and the data …), but I won't get such a machine in the near future. Anyone reading these lines in possession of a recent Apple silicon device, please speak up, and we'll happily try to instruct you in how to get the data we need!

But even without the Apple silicon part, I think it would be interesting to have some sort of benchmark numbers included in the repository, for people to see whether their system lives up to the "standard" numbers. @ankrh, we could think of which file(s) might make sense to run for how long for such a test. I agree with Sacroth that we might want to run both short simulations (for users to get a quick idea about the performance of their systems) and long ones to see whether there are any thermal limits. What is your take on this?

**Reply by ankrh** [reply]
It could be neat to show the performance (photons/min) people have gotten on different hardware, but I'm not sure what would be the most elegant way to provide that here on GitHub. Maybe an open discussions thread where everyone can just add a comment with their result?

For the specific model to use for benchmarking, I don't think it matters too much as long as it's a model with a decent about of scattering and a decent amount of absorption. As you suggest, example 16 would be OK for this. In its current form it runs for a short time (6 sec for the MC part, 1½ minutes for the heat sim part). For a long test we can set the MC to 30 mins and nPulses to 200.
